File: leetcode/716-max-stack/main.go
package main

import "fmt"

/*
	Common approach
	- store the value and max for each item so the push(), pop(), top(), peek() are O(1)
	- but PopMax() inevitably needs to find top-most max item and update the top items' max, it takes O(n)
	60ms beats 57.14%
	26jan2019
*/

type Item struct {
	Val int
	Max int
}

type MaxStack struct {
	Stack []Item
}

/** initialize your data structure here. */
func Constructor() MaxStack {
	return MaxStack{[]Item{}}
}

func (this *MaxStack) Push(x int) {
	if len(this.Stack) == 0 {
		this.Stack = append(this.Stack, Item{x, x})
		return
	}
	max := this.PeekMax()
	if x > max {
		this.Stack = append(this.Stack, Item{x, x})
	} else {
		this.Stack = append(this.Stack, Item{x, max})
	}
}

func (this *MaxStack) Pop() int {
	if len(this.Stack) == 0 {
		return -1
	}
	temp := this.Stack[len(this.Stack)-1]
	this.Stack = this.Stack[:len(this.Stack)-1]
	return temp.Val
}

func (this *MaxStack) Top() int {
	if len(this.Stack) == 0 {
		return -1
	}
	return this.Stack[len(this.Stack)-1].Val
}

func (this *MaxStack) PeekMax() int {
	if len(this.Stack) == 0 {
		return -1
	}
	return this.Stack[len(this.Stack)-1].Max
}

func (this *MaxStack) PopMax() int {
	if len(this.Stack) == 0 {
		return -1
	}
	// rmeove max
	max := this.PeekMax()
	var temp Item
	i := len(this.Stack) - 1
	for ; i >= 0; i-- {
		if max == this.Stack[i].Val {
			temp = this.Stack[i]
			clone := []Item{}
			clone = append(clone, this.Stack[:i]...)
			clone = append(clone, this.Stack[i+1:]...)
			this.Stack = clone
			break
		}
	}

	// update items' max
	for ; i < len(this.Stack); i++ {
		cur := this.Stack[i]
		if i > 0 {
			prev := this.Stack[i-1]
			if cur.Val > prev.Max {
				this.Stack[i].Max = cur.Val
			} else {
				this.Stack[i].Max = prev.Max
			}
		} else {
			this.Stack[i].Max = cur.Val
		}
	}

	return temp.Val
}

func main() {
	// s := Constructor()
	// s.Push(5)
	// s.Push(1)
	// s.Push(5)
	// fmt.Println(s.Top())     // 5
	// fmt.Println(s.PopMax())  // pop 5
	// fmt.Println(s.Top())     // 1
	// fmt.Println(s.PeekMax()) // 5
	// fmt.Println(s.Pop())     // pop 1
	// fmt.Println(s.Top())     // 5

	s := Constructor()
	s.Push(5)
	s.Push(1)
	fmt.Println(s.PopMax())
	fmt.Println(s.PeekMax())
	fmt.Println(s.PopMax())
	fmt.Println(s.Stack)
}

File: leetcode/716-max-stack/main.py
"""
    1st approach: doubly linked list + binary search

    Time of
    - push()      O(logN + K)
    - pop()       O(logN + K)
    - top()       O(1)
    - peekMax()   O(1)
    - popMax()    O(logN + K)
    256 ms, faster than 17.40%
"""


class DLLNode(object):
    def __init__(self, val):
        self.val = val
        self.next = None
        self.prev = None


class MaxStack(object):

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.nums = []
        self.head = DLLNode(0)
        self.tail = DLLNode(0)
        self.head.next = self.tail
        self.tail.prev = self.head

    def push(self, x):
        """
        :type x: int
        :rtype: None
        """
        last = self.tail.prev
        node = DLLNode(x)
        last.next = node
        node.prev = last
        node.next = self.tail
        self.tail.prev = node

        idx = self.upper_bsearch(self.nums, x)
        self.nums.insert(idx, (x, node))

    def pop(self):
        """
        :rtype: int
        """
        last = self.tail.prev
        last.prev.next = last.next
        last.next.prev = last.prev

        idx = self.upper_bsearch(self.nums, last.val)
        self.nums.pop(idx-1)
        return last.val

    def top(self):
        """
        :rtype: int
        """
        return self.tail.prev.val

    def peekMax(self):
        """
        :rtype: int
        """
        return self.nums[-1][0]

    def popMax(self):
        """
        :rtype: int
        """
        maxVal, maxNode = self.nums[-1]
        maxNode.prev.next = maxNode.next
        maxNode.next.prev = maxNode.prev

        idx = self.upper_bsearch(self.nums, maxVal)
        self.nums.pop(idx-1)
        return maxVal

    def upper_bsearch(self, nums, target):
        left = 0
        right = len(nums)
        while left < right:
            mid = (left + right)/2
            if target >= nums[mid][0]:
                left = mid + 1
            else:
                right = mid
        return left


"""
    Easier approach: sorted list + stack

    Time of
    - push()      O(logN + K)
    - pop()       O(logN + K)
    - top()       O(1)
    - peekMax()   O(1)
    - popMax()    O(K)
    148 ms, faster than 80.07%
"""


class MaxStack:

    def __init__(self):
        self.sortedNums = []
        self.stack = []

    def push(self, x: int) -> None:
        self.stack.append(x)
        i = self._upperBsearch(self.sortedNums, x)
        self.sortedNums.insert(i, x)

    def pop(self) -> int:
        res = self.stack.pop()
        i = self._upperBsearch(self.sortedNums, res)
        self.sortedNums.pop(i-1)
        return res

    def top(self) -> int:
        return self.stack[-1]

    def peekMax(self) -> int:
        return self.sortedNums[-1]

    def popMax(self) -> int:
        res = self.sortedNums.pop()
        for i in range(len(self.stack)-1, -1, -1):
            if self.stack[i] == res:
                self.stack.pop(i)
                break
        return res

    def _upperBsearch(self, nums, target):
        left = 0
        right = len(nums)
        while left < right:
            mid = (left + right)//2
            if target >= nums[mid]:
                left = mid + 1
            else:
                right = mid
        return right
